File: kotlin/kotlin/pl/zarajczyk/my_family/RunningApplicationsFetcher.kt
package pl.zarajczyk.my_family

import com.lordcodes.turtle.shellRun
import org.apache.commons.lang3.SystemUtils
import org.springframework.stereotype.Service

@Service
class RunningApplicationsFetcher {
    fun getRunningApps(): List<String> = when {
        SystemUtils.IS_OS_MAC -> getRunningAppsMacOs()
        else -> throw RuntimeException("Unsupported operating system")
    }

    private fun getRunningAppsMacOs(): List<String> {
        val user = SystemUtils.getUserName()
        return shellRun("ps", listOf("-u $user", "-o comm"))
            .lines()
            .filter { it.startsWith("/Applications/") }
            .map { it.substring(0, it.indexOf(".app/") + 4) }
            .map { it.removePrefix("/Applications/") }
            .distinct()
            .toList()
    }
}

File: src/RunningApplicationsFetcher.py
import subprocess

from osutils import is_mac


class RunningApplicationsFetcher:
    def get_running_apps(self):
        if is_mac():
            return self.__get_running_apps_mac_os()
        else:
            raise Exception("Unsupported operating system")

    def __get_running_apps_mac_os(self):
        user = subprocess.run("whoami", capture_output=True, text=True).stdout.strip()
        result = subprocess.run(["ps", "-u", user, "-o", "comm"], capture_output=True, text=True)
        lines = result.stdout.splitlines()
        apps = [line for line in lines if line.startswith("/Applications/")]
        apps = [app[:app.index(".app/") + 4] for app in apps]
        apps = [app.removeprefix("/Applications/") for app in apps]
        return list(set(apps))
